In [1]:
!pip install ultralytics

In [2]:
from ultralytics import YOLO

model = YOLO("yolo11n-pose.pt")

print("Model Loaded")

Model Loaded


In [3]:
from google.colab import files

uploaded = files.upload()

Saving 7187424-hd_1920_1080_24fps.mp4 to 7187424-hd_1920_1080_24fps (4).mp4


In [4]:
!ls

'7187424-hd_1920_1080_24fps (1).mp4'   7187424-hd_1920_1080_24fps.mp4
'7187424-hd_1920_1080_24fps (2).mp4'   runs
'7187424-hd_1920_1080_24fps (3).mp4'   sample_data
'7187424-hd_1920_1080_24fps (4).mp4'   yolo11n-pose.pt


In [5]:
results = model.predict(
    source="7187424-hd_1920_1080_24fps.mp4",
    save=True,
    conf=0.5
)


WARNING ⚠️ 
Inference results will accumulate in RAM unless `stream=True` is passed, which can cause out-of-memory errors for large
sources or long-running streams and videos. See https://docs.ultralytics.com/modes/predict/ for help.

Example:
    results = model(source=..., stream=True)  # generator of Results objects
    for r in results:
        boxes = r.boxes  # Boxes object for bbox outputs
        masks = r.masks  # Masks object for segment masks outputs
        probs = r.probs  # Class probabilities for classification outputs

video 1/1 (frame 1/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 182.6ms
video 1/1 (frame 2/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 160.3ms
video 1/1 (frame 3/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 160.8ms
video 1/1 (frame 4/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 155.8ms
video 1/1 (frame 5/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 166.8ms
vid

In [6]:
!ls runs/pose/predict

7187424-hd_1920_1080_24fps.avi


In [7]:
from IPython.display import Video

Video(
    "runs/pose/predict/7187424-hd_1920_1080_24fps.mp4",
    embed=True
)

In [8]:
!find runs -type f

runs/pose/predict-3/7187424-hd_1920_1080_24fps.avi
runs/pose/predict-4/7187424-hd_1920_1080_24fps.avi
runs/pose/predict/7187424-hd_1920_1080_24fps.avi
runs/pose/predict-2/7187424-hd_1920_1080_24fps.avi


In [9]:
!ls -lh runs/pose/predict

total 20M
-rw-r--r-- 1 root root 20M Jun  5 04:34 7187424-hd_1920_1080_24fps.avi


In [10]:
results = model.predict(
    source="7187424-hd_1920_1080_24fps.mp4",
    save=False
)

print(type(results))
print(len(results))


WARNING ⚠️ 
Inference results will accumulate in RAM unless `stream=True` is passed, which can cause out-of-memory errors for large
sources or long-running streams and videos. See https://docs.ultralytics.com/modes/predict/ for help.

Example:
    results = model(source=..., stream=True)  # generator of Results objects
    for r in results:
        boxes = r.boxes  # Boxes object for bbox outputs
        masks = r.masks  # Masks object for segment masks outputs
        probs = r.probs  # Class probabilities for classification outputs

video 1/1 (frame 1/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 148.7ms
video 1/1 (frame 2/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 156.6ms
video 1/1 (frame 3/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 155.8ms
video 1/1 (frame 4/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 171.1ms
video 1/1 (frame 5/168) /content/7187424-hd_1920_1080_24fps.mp4: 384x640 1 person, 147.6ms
vid

In [11]:
r = results[0]

print(r.keypoints)

ultralytics.engine.results.Keypoints object with attributes:

conf: tensor([[7.9037e-01, 7.5699e-01, 3.5788e-01, 8.7019e-01, 7.4944e-02, 8.7331e-01, 9.6700e-01, 1.5554e-01, 5.7832e-01, 1.5978e-01, 3.8656e-01, 2.0640e-02, 5.1263e-02, 6.5971e-04, 1.6954e-03, 1.5344e-04, 3.1886e-04]])
data: tensor([[[9.9320e+02, 3.9949e+01, 7.9037e-01],
         [1.0976e+03, 0.0000e+00, 7.5699e-01],
         [9.6440e+02, 0.0000e+00, 3.5788e-01],
         [1.3029e+03, 5.9419e+01, 8.7019e-01],
         [9.6300e+02, 5.0188e+01, 7.4944e-02],
         [1.6273e+03, 5.4653e+02, 8.7331e-01],
         [7.5528e+02, 5.2604e+02, 9.6700e-01],
         [1.7746e+03, 1.0800e+03, 1.5554e-01],
         [4.4865e+02, 1.0187e+03, 5.7832e-01],
         [1.5630e+03, 1.0356e+03, 1.5978e-01],
         [5.0340e+02, 9.7620e+02, 3.8656e-01],
         [1.3635e+03, 1.0800e+03, 2.0640e-02],
         [8.0464e+02, 1.0800e+03, 5.1263e-02],
         [1.2611e+03, 9.7446e+02, 6.5971e-04],
         [6.3514e+02, 9.4882e+02, 1.6954e-03],
      

In [12]:
print(r.keypoints.xy.shape)

torch.Size([1, 17, 2])


1 person,

17 body keypoints,

(x,y) coordinates

In [13]:
print(r.keypoints.xy[0])

tensor([[ 993.2003,   39.9493],
        [1097.6060,    0.0000],
        [ 964.3990,    0.0000],
        [1302.8551,   59.4190],
        [ 962.9977,   50.1881],
        [1627.2898,  546.5283],
        [ 755.2766,  526.0444],
        [1774.6234, 1080.0000],
        [ 448.6470, 1018.7353],
        [1562.9672, 1035.6111],
        [ 503.4044,  976.2040],
        [1363.4648, 1080.0000],
        [ 804.6365, 1080.0000],
        [1261.0540,  974.4625],
        [ 635.1371,  948.8199],
        [1034.9188, 1042.5824],
        [ 620.7162, 1041.5559]])


In [14]:
print(r.keypoints.xy.shape)

torch.Size([1, 17, 2])


In [15]:
print(r.keypoints.xy[0])

tensor([[ 993.2003,   39.9493],
        [1097.6060,    0.0000],
        [ 964.3990,    0.0000],
        [1302.8551,   59.4190],
        [ 962.9977,   50.1881],
        [1627.2898,  546.5283],
        [ 755.2766,  526.0444],
        [1774.6234, 1080.0000],
        [ 448.6470, 1018.7353],
        [1562.9672, 1035.6111],
        [ 503.4044,  976.2040],
        [1363.4648, 1080.0000],
        [ 804.6365, 1080.0000],
        [1261.0540,  974.4625],
        [ 635.1371,  948.8199],
        [1034.9188, 1042.5824],
        [ 620.7162, 1041.5559]])


In [16]:
#Extract Shoulder, Elbow, Wrist

import numpy as np

pts = r.keypoints.xy[0].cpu().numpy()

right_shoulder = pts[6]
right_elbow    = pts[8]
right_wrist    = pts[10]

print("Shoulder:", right_shoulder)
print("Elbow:", right_elbow)
print("Wrist:", right_wrist)

Shoulder: [     755.28      526.04]
Elbow: [     448.65      1018.7]
Wrist: [      503.4       976.2]


In [17]:
#Angle Function

import numpy as np

def calculate_angle(a, b, c):

    a = np.array(a)
    b = np.array(b)
    c = np.array(c)

    radians = np.arctan2(
        c[1]-b[1],
        c[0]-b[0]
    ) - np.arctan2(
        a[1]-b[1],
        a[0]-b[0]
    )

    angle = np.abs(radians * 180 / np.pi)

    if angle > 180:
        angle = 360 - angle

    return angle

In [18]:
#Calculate Elbow Angle

angle = calculate_angle(
    right_shoulder,
    right_elbow,
    right_wrist
)

print("Elbow Angle =", angle)

Elbow Angle = 20.266317


Elbow Angle = 20.27°

This means the arm is fully bent, exactly what we expect at the top of a dumbbell curl.

160° - 180°  → Arm straight

20° - 40°    → Arm curled

So your angle calculation is working correctly

In [19]:
#Count Repetitions

import cv2

cap = cv2.VideoCapture(
    "7187424-hd_1920_1080_24fps.mp4"
)

frame_count = 0

while True:

    ret, frame = cap.read()

    if not ret:
        break

    frame_count += 1

cap.release()

print("Frames =", frame_count)

Frames = 168


Frames = 168,

Elbow Angle = 20.27°

In [20]:
#Person
#Holding dumbbell
#Performing bicep curls

counter = 0
stage = None

if angle > 160:
    stage = "down"

if angle < 40 and stage == "down":
    stage = "up"
    counter += 1

Straight Arm
   ↓
Bent Arm
   ↓
Straight Arm

= 1 Rep

In [21]:
#Upload a real dumbbell curl video

from google.colab import files

uploaded = files.upload()

Saving 7187424-hd_1920_1080_24fps.mp4 to 7187424-hd_1920_1080_24fps (5).mp4


In [22]:
!ls

'7187424-hd_1920_1080_24fps (1).mp4'   7187424-hd_1920_1080_24fps.mp4
'7187424-hd_1920_1080_24fps (2).mp4'   runs
'7187424-hd_1920_1080_24fps (3).mp4'   sample_data
'7187424-hd_1920_1080_24fps (4).mp4'   yolo11n-pose.pt
'7187424-hd_1920_1080_24fps (5).mp4'


In [23]:
counter = 0
stage = None

angles = []

In [24]:
#Create Angle Function

import numpy as np

def calculate_angle(a, b, c):
    a = np.array(a)
    b = np.array(b)
    c = np.array(c)

    radians = np.arctan2(c[1]-b[1], c[0]-b[0]) - \
              np.arctan2(a[1]-b[1], a[0]-b[0])

    angle = np.abs(radians * 180.0 / np.pi)

    if angle > 180:
        angle = 360 - angle

    return angle


In [25]:
#Process Entire Video

import cv2
from ultralytics import YOLO

model = YOLO("yolo11n-pose.pt")

video_path = "7187424-hd_1920_1080_24fps.mp4"

cap = cv2.VideoCapture(video_path)

counter = 0
stage = None

frame_no = 0

while True:

    ret, frame = cap.read()

    if not ret:
        break

    frame_no += 1

    # Process every 3rd frame (faster)
    if frame_no % 3 != 0:
        continue

    results = model(frame, verbose=False)

    if len(results) == 0:
        continue

    if results[0].keypoints is None:
        continue

    pts = results[0].keypoints.xy

    if len(pts) == 0:
        continue

    pts = pts[0].cpu().numpy()

    right_shoulder = pts[6]
    right_elbow = pts[8]
    right_wrist = pts[10]

    angle = calculate_angle(
        right_shoulder,
        right_elbow,
        right_wrist
    )

    # Curl logic
    if angle > 150:
        stage = "down"

    if angle < 50 and stage == "down":
        stage = "up"
        counter += 1

        print("Rep:", counter)

cap.release()

print("\nTotal Repetitions =", counter)

Rep: 1
Rep: 2
Rep: 3
Rep: 4
Rep: 5

Total Repetitions = 5


Video
  ↓
YOLO Pose
  ↓
Body Keypoints
  ↓
Elbow Angle
  ↓
Rep Counter
  ↓
Total Repetitions

In [26]:
#Create Gradio Web App

import gradio as gr
import cv2
import numpy as np
from ultralytics import YOLO

model = YOLO("yolo11n-pose.pt")

def calculate_angle(a, b, c):

    a = np.array(a)
    b = np.array(b)
    c = np.array(c)

    radians = np.arctan2(c[1]-b[1], c[0]-b[0]) - \
              np.arctan2(a[1]-b[1], a[0]-b[0])

    angle = np.abs(radians * 180 / np.pi)

    if angle > 180:
        angle = 360 - angle

    return angle


def count_reps(video_path):

    cap = cv2.VideoCapture(video_path)

    counter = 0
    stage = None

    frame_no = 0

    while True:

        ret, frame = cap.read()

        if not ret:
            break

        frame_no += 1

        if frame_no % 3 != 0:
            continue

        results = model(frame, verbose=False)

        if len(results) == 0:
            continue

        if results[0].keypoints is None:
            continue

        pts = results[0].keypoints.xy

        if len(pts) == 0:
            continue

        pts = pts[0].cpu().numpy()

        shoulder = pts[6]
        elbow = pts[8]
        wrist = pts[10]

        angle = calculate_angle(
            shoulder,
            elbow,
            wrist
        )

        if angle > 150:
            stage = "down"

        if angle < 50 and stage == "down":
            stage = "up"
            counter += 1

    cap.release()

    return f"Total Repetitions: {counter}"

In [27]:
#Launch Gradio

demo = gr.Interface(
    fn=count_reps,
    inputs=gr.Video(),
    outputs="text",
    title="Dumbbell Curl Counter AI"
)

demo.launch(share=True)

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://fd5c263da7365814c4.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)
